File: neetcode150/graphs/leetcode261.py
from typing import List

from utilities.union_find import UnionFind


class ValidTreeUnionFind(UnionFind):
    def introduces_cycle(self, u: int, v: int) -> bool:
        pu = self.find(u)
        pv = self.find(v)

        if pu == pv:
            return True

        if self.rank[pu] < self.rank[pv]:
            self.rank[pv] += 1
            self.parent[pu] = pv
        else:
            self.rank[pu] += 1
            self.parent[pv] = pu

        return False


class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        """
        You have a graph of n nodes labeled from 0 to n - 1.
        You are given an integer n and a list of edges where edges[i] = [ai, bi]
        indicates that there is an undirected edge between nodes ai and bi in the graph.

        Return true if the edges of the given graph make up a valid tree, and false otherwise.

        Approach:

        We know if it's a valid tree if there is no cycle.

        using union find, we know that an edge uv introduces a cycle, if u and v already share
        the same parent node

        :param n:
        :param edges:
        :return:
        """
        uf = ValidTreeUnionFind(n)

        for u, v in edges:
            # if the edge introduces a cycle, we know it's not a valid tree
            if uf.introduces_cycle(u, v):
                return False

        # the last check we need to do is make sure that we only have 1
        # connected component
        return uf.num_connected_components == 1

File: js_algorithms/neetcode150/graphs/leetcode261.ts
class UnionFind {
  parents: Array<number>;
  ranks: Array<number>;

  constructor(n: number) {
    this.parents = new Array(n).fill(0).map((_, index) => index);
    this.ranks = new Array(n).fill(0);
  }

  find(u: number): number {
    while (u !== this.parents[u]) {
      this.parents[u] = this.parents[this.parents[u]];
      u = this.parents[u];
    }

    return u;
  }

  union(u: number, v: number) {
    const pu = this.find(u);
    const pv = this.find(v);

    if (pu === pv) {
      return;
    }

    if (this.ranks[pu] < this.ranks[pv]) {
      this.parents[pu] = pv;
      this.ranks[pv]++;
    } else {
      this.parents[pv] = pu;
      this.ranks[pu]++;
    }
  }
}

function validTree(n: number, edges: number[][]): boolean {
  // if the nodes of an incoming edge have the same parent
  // this means that we can form a cycle

  const uf = new UnionFind(n);

  for (const [u, v] of edges) {
    const pu = uf.find(u);
    const pv = uf.find(v);

    if (pu === pv) {
      return false;
    }

    uf.union(u, v);
  }

  // we also have to check how many connected components there are
  const connectedComponents: Set<number> = new Set();
  for (let i = 0; i < n; i++) {
    connectedComponents.add(uf.find(i));
  }

  return connectedComponents.size === 1;
}

export default validTree;
